### PR TITLE
[Model Zoo] GCN on Tox21 Dataset

### PR DESCRIPTION
**Pull request by mufeili**
## Description
You can start review with `model_zoo/drug_discovery/README.md` and then `model_zoo/drug_discovery/gcn_for_property_prediction/README.md`.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [x] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR

### CODE REVIEW

File: docs/source/api/python/data.rst [hunk 1]
```diff
@@ -57,3 +57,14 @@ Protein-Protein Interaction dataset
 
 .. autoclass:: PPIDataset
     :members: __getitem__, __len__
+
+Molecular Graphs
+````````````````
+
+To work on molecular graphs, make sure you have installed `RDKit <https://www.rdkit.org/docs/Install.html>`__.
+
+.. autofunction:: dgl.data.molecule.default_atom_featurizer
```
Comment (VoVAllen) [inline docs/source/api/python/data.rst]: `dgl.data.molecule.atom_featurizer(f=default_one)`?

File: docs/source/api/python/data.rst [hunk 2]
```diff
@@ -57,3 +57,14 @@ Protein-Protein Interaction dataset
 
 .. autoclass:: PPIDataset
     :members: __getitem__, __len__
+
+Molecular Graphs
+````````````````
+
+To work on molecular graphs, make sure you have installed `RDKit <https://www.rdkit.org/docs/Install.html>`__.
```
Comment (VoVAllen) [inline docs/source/api/python/data.rst]: Add suggested version here also.

File: model_zoo/drug_discovery/README.md [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+mostly developed based on molecule fingerprints.
+
+Graph neural networks make it possible for a data-driven representation of molecules out of the atoms, bonds and 
+molecular graph topology, which may be viewed as a learned fingerprint [3]. For details, checkout the README of 
+`gcn for property prediction`.
+
+### Models
+- **Graph Convolutional Networks**: Graph Convolutional Networks (GCN) have been one of the most popular graph neural 
+network and they can be easily extended for graph level prediction.
+
```
Comment (VoVAllen) [inline model_zoo/drug_discovery/README.md]: Will I add generative model here? Or another file?

File: model_zoo/drug_discovery/gcn_for_property_prediction/README.md [hunk 4]
```diff
@@ -0,0 +1,53 @@
+# Graph Convolutional Networks for Molecule Property Prediction
+
+We consider this model to be the first step stone for drug discovery with DGL. If you have used DeepChem before, you
+should find our model very similar to 
+[DeepChem's tutorial on graph convolution](https://deepchem.io/docs/notebooks/graph_convolutional_networks_for_tox21.html).
```
Comment (VoVAllen) [inline model_zoo/drug_discovery/gcn_for_property_prediction/README.md]: Do we need to say this?

File: model_zoo/drug_discovery/gcn_for_property_prediction/main.py [hunk 5]
```diff
@@ -0,0 +1,315 @@
...
+def load_checkpoint(model, checkpoint_path, load_optimizer, optimizer=None):
+    """Load checkpoint for model (and optimizer).
+
+    Parameters
+    ----------
+    model
+        Model to load checkpoint
+    checkpoint_path : str
+        Path to saved checkpoint
+    load_optimizer : bool
```
Comment (VoVAllen) [inline model_zoo/drug_discovery/gcn_for_property_prediction/main.py]: load_optimizer flag is redundant.

File: python/dgl/data/molecule.py [hunk 6]
```diff
@@ -0,0 +1,471 @@
...
+        ----------
+        frac_train : float
+            Proportion of data to use for training
+        frac_val : float
+            Proportion of data to use for validation
+        frac_test : float
+            Proportion of data to use for test
+        """
+        frac_total = frac_train + frac_val + frac_test
+        assert frac_total == 1, \
```
Comment (VoVAllen) [inline python/dgl/data/molecule.py]: adding floats may have precision problem. Also I think we might need a BaseObject for all molecule dataset since there're multiple datasets for similar tasks.

File: python/dgl/data/molecule.py [hunk 7]
```diff
@@ -0,0 +1,471 @@
...
+        Returns
+        -------
+        DGLGraph
+            DGLGraph of the datapoint
+        Tensor of dtype float32
+            Labels of the datapoint for all tasks
+        Tensor of dtype float32
+            Weights of the datapoint for all tasks
+        """
+        return self.graphs[item], self.labels[item], self.weights[item]
```
Comment (VoVAllen) [inline python/dgl/data/molecule.py]: Personally I think we should follow the design of pytorch dataloader/dataset. A `Dataset` implements `_getitem_`, `torch.utils.data.Subset(dataset, indices)` to get train/test/val set, and `WeightedRandomSampler` if we need to rebalance the data.

File: model_zoo/drug_discovery/README.md (hunk 3 above)
Comment (mufeili) [reply model_zoo/drug_discovery/README.md]: I would suggest using this README and we can add multiple sections for different task categories.

File: docs/source/api/python/data.rst (hunk 2 above)
[the remark this answers is not in the document]
Comment (mufeili) [reply docs/source/api/python/data.rst]: Could you please leave instructions for installing version `2018.09.3`?

File: docs/source/api/python/data.rst (hunk 1 above)
Comment (mufeili) [reply docs/source/api/python/data.rst]: I'm now directly passing an `atom_featurizer` to `mol2dgl`. Think this like a user defined message function. I think we probably don't need a two-level wrapping here.

File: model_zoo/drug_discovery/gcn_for_property_prediction/README.md (hunk 4 above)
Comment (mufeili) [reply model_zoo/drug_discovery/gcn_for_property_prediction/README.md]: One goal of this model is to replicate DeepChem's graph convolution on Tox21 as a GNN 101 for drug discovery. As a significant proportion of our potential audience might have used DeepChem, I think this comment might be helpful.

File: docs/source/api/python/data.rst (hunk 2 above)
Comment (mufeili) [reply docs/source/api/python/data.rst]: Seems that you commented at one place twice?

File: model_zoo/drug_discovery/gcn_for_property_prediction/main.py (hunk 5 above)
Comment (mufeili) [reply model_zoo/drug_discovery/gcn_for_property_prediction/main.py]: Done.

File: docs/source/api/python/data.rst (hunk 2 above)
[the remark this answers is not in the document]
Comment (mufeili) [reply docs/source/api/python/data.rst]: I confirmed that this model works with `2018.09.3`.

File: python/dgl/data/molecule.py (hunk 6 above)
Comment (mufeili) [reply python/dgl/data/molecule.py]: This is now replaced with `np.allclose(frac_total, 1.)`.

File: python/dgl/data/molecule.py (hunk 7 above)
Comment (mufeili) [reply python/dgl/data/molecule.py]: I've adapted PyTorch's `Subset`, but the training/validation/test set are still split during initialization as dataset splits can be very different with molecules.

**Comment by mufeili** [comment]
#760 refactored the code for the dataset.
